### DISCUSSION TITLE
How to make multiple output of an asset without pre-define outs?

### DISCUSSION THREAD

**Discussion by meyergin** [dagster-io/dagster#15698] [Q&A] [answered]
is it possible to write an asset  which fetches a list of data and yields multiple outputs to save data seperately.
In the tutorial, the multi-asset must pre-define outs and dynamical graph seems use for map-reduce instead of for-loop sequentially.
Or should I just use op to for-loop save files instead?
scenario: use api to fetch data and save it by the unique key.
op pseudo-code:
```
def op():
  for key in big_lists:
    data = _get_data(key)
    # save data to the file named by key
    save(data,key)
```

**Comment by OwenKephart** [comment, marked as answer]
Hi @meyergin -- there's no explicit requirement that the IOManager of an asset handles storing the data, so it'd be acceptable to write:

```python
@asset
def my_asset() -> None:
    for key in big_lists:
        data = _get_data(key)
        save(data,key)
```

This ends up just having a single (implicit) output, but writes data separately.

The only issue here is if you have a downstream asset, it might be more difficult to figure out what exact data was stored in the above step. You could also imagine writing a fancy IOManager to do this for you, but that might be more trouble than it's worth for your usecase: https://github.com/dagster-io/dagster/discussions/9772
